[libc++][span] P2447R4: `std::span` over an initializer list

H-G-Hristov opened the pull request:
Implements: https://wg21.link/P2447R6
- https://eel.is/c++draft/span.syn
- https://eel.is/c++draft/span.overview
- https://eel.is/c++draft/span.cons
- https://eel.is/c++draft/diff

Zingam committed:
[libc++][span] P2447R4: `std::span` over an initializer list

Implements: https://wg21.link/P2447R6
- https://eel.is/c++draft/span.syn
- https://eel.is/c++draft/span.overview
- https://eel.is/c++draft/span.cons
- https://eel.is/c++draft/diff

Zingam committed:
WIP: Cleanup

Zingam committed:
Added generated version files

Zingam committed:
Added SFINAE

Zingam committed:
Cleanup

Zingam committed:
Try to fix CI

Zingam committed:
Try to fix CI

mordante commented on libcxx/test/std/containers/views/views.span/span.cons/iterator_len.verify.cpp:
```diff
@@ -13,22 +13,36 @@
-  std::span<int, 2> sp2 = {0, 0}; // expected-error {{no matching constructor for initialization of 'std::span<int, 2>'}}
-  std::span<const int> csp = {0, 0}; // expected-error {{no matching constructor for initialization of 'std::span<const int>'}}
-  std::span<const int, 2> csp2 = {0, 0}; // expected-error {{no matching constructor for initialization of 'std::span<const int, 2>'}}
+  // expected-error@+1 {{no matching constructor for initialization of 'std::span<int>'}}
```
Can you add tests for verify all changes listed in Annex C?

mordante commented on libcxx/include/span:
```diff
@@ -228,6 +229,14 @@ public:
+  _LIBCPP_HIDE_FROM_ABI constexpr explicit span(std::initializer_list<value_type> __il)
+    requires is_const_v<element_type>
+      : __data_{__il.begin()} {
+    _LIBCPP_ASSERT_INTERNAL(_Extent == __il.size(), "Size mismatch in span's constructor _Extent != __il.size().");
```
```suggestion
    _LIBCPP_ASSERT_VALID_INPUT_RANGE(_Extent == __il.size(), "Size mismatch in span's constructor _Extent != __il.size().");
```

mordante commented on libcxx/test/std/containers/views/views.span/span.cons/iterator_len.verify.cpp:
```diff
@@ -13,22 +13,36 @@
+#if TEST_STD_VER >= 26
+  // No error in C++26
+  std::span<const int> csp = {0, 0};
+  // expected-error@+1 {{chosen constructor is explicit in copy-initialization}}
```
We typically test explicitness without a verify test.
The main function in `test/std/input.output/file.streams/fstreams/ifstream.cons/path.pass.cpp`.

In general we prefer to test things without verify tests. These are only tested by us not by other standard libraries using our test suite. They are sensitive to changes in Clang diagnostics.

H-G-Hristov replied:
Thanks! I'll update the tests.

mordante reviewed (changes requested):
Mostly looks good, some minor issues.

Zingam committed:
WIP: Addressed some comments

Zingam committed:
Addressed comments

H-G-Hristov commented:
Thanks. The documentation build fails due to some CMake error consistently otherwise I believe I addressed all comments.

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2447R4-span-over-initializer_list

mordante commented on libcxx/test/std/containers/views/views.span/span.cons/initializer_list.assert.pass.cpp:
```diff
@@ -0,0 +1,38 @@
+}
+
+int main(int, char**) {
+  assert(test());
```
We typically don't do this for run-time tests. We do it for compile-time tests so we can use `static_assert`s. In this case I would remove the `test` function and move the test code in `main`.

I overlooked this in the previous review.

mordante commented on libcxx/test/std/containers/views/views.span/span.cons/initializer_list.pass.cpp:
```diff
@@ -5,39 +5,134 @@
+constexpr int four(std::span<const std::any> sp) { return sp.size(); }
+
+bool test_P2447R4_annex_c_examples() {
+  // 1. Overload resolution is affected
+  // --> tested in "initializer_list.verify.cpp"
+
+  // 2. The `initializer_list` ctor has high precedence
+  // --> tested in "initializer_list.verify.cpp"
+
+  // 3. Implicit two-argument construction with a highly convertible value_type
```
nice! thanks for referring to the other test too!

mordante reviewed (approved):
> Thanks. The documentation build fails due to some CMake error consistently otherwise I believe I addressed all comments.
I think I fixed this a few days ago.

LGTM modulo 1 nit.

H-G-Hristov committed:
Merge branch 'main' into hgh/libcxx/P2447R4-span-over-initializer_list

Zingam committed:
Addressed comment

H-G-Hristov commented:
> > Thanks. The documentation build fails due to some CMake error consistently otherwise I believe I addressed all comments.
> > I think I fixed this a few days ago.
> 
> LGTM modulo 1 nit.

Thanx!
